class.c

```diff
@@ -979,11 +988,27 @@ include_modules_at(const VALUE klass, VALUE c, VALUE module, int search_super)
 	iclass = rb_include_class_new(module, super_class);
 	c = RCLASS_SET_SUPER(c, iclass);
         RCLASS_SET_INCLUDER(iclass, klass);
+        add_subclass = TRUE;
+        if (module != RCLASS_ORIGIN(module)) {
+            tmp_stack = orig_stack;
+            orig_stack = malloc(sizeof(struct origin_stack));
```

Comment:
Can't this cause potential memory leak, when

Reply:
This pointer is freed later when the origin iclass is created.  However, if there is an exception raised between when this pointer is created and when it is freed, then that would cause a small memory leak (12-24 bytes depending on arch * number of items in stack).  I do not know if any of the functions in the`while` loop can raise an exception, but it's complex enough that I cannot rule it out. Should I switch this to use `rb_alloc_tmp_buffer2`?

Reply:
I figured out a better approach, by storing the information in a hidden Ruby array.  This should only increase memory by 8-16 bytes per module origin iclass.  Assuming it passes CI, I'll commit it.

Reply:
In this loop, `rb_include_class_new`, `rb_module_add_to_subclasses_list`, `rb_module_add_to_subclasses_list` and `add_refined_method_entry_i` can cause `NoMemoryError`.
How about pushing iclass and origin to a hidden array?

Reply:
Already implemented in the latest commit. :)

Reply:
I mean pushing to a _single_ array instead of chaining.

```c
        add_subclass = TRUE;
        if (module != RCLASS_ORIGIN(module)) {
            if (!origin_stack) origin_stack = rb_ary_tmp_new(2);
            VALUE origin[2] = {iclass, RCLASS_ORIGIN(module)};
            rb_ary_cat(2, origin, origin_stack);
        }
        else if (origin_stack && (origin_len = RARRAY_LEN(ary)) > 1 &&
                 RARRAY_AREF(origin_stack, origin_len - 1) == module) {
            RCLASS_SET_ORIGIN(RARRAY_AREF(origin_stack, (origin_len -= 2)), iclass);
            rb_ary_resize(origin_stack, origin_len);
            add_subclass = FALSE;
        }
```

Reply:
Sure, I'll push another commit that uses that approach.